### PR TITLE
KeyConfig reload on CS-Go

### PR DESCRIPTION
**Pull request by Oncilla**
Enable KeyConfig reload on CS-Go. The reload is started by sending SIG_HUP to the certificate server.
Following is reloaded:
- Trust store
- KeyConfig

The keys are loaded from gen/keys. The keys in KeyConfig are only replaced, if the new keys are usable with the trust files.

Root keys are usable if:
- They are authenticated by an active TRC (newest or in grace period)

Signing keys are usable if:
 - They are authenticated by the newest certificate chain (*)
 - They authenticating certificate chain is verifiable with active TRC (newest or in grace period)

(*) This is necessary since the core AS issuing leaf certificates has a mapping [AS->verifying key] and will only accept reissuance requests if the key matches. 
Thus having a different key (even if authenticated by an older but still active certificate chain) means the certificate server is not fully operational.

---
This change is [Reviewable](https://reviewable.io/reviews/scionproto/scion/1377)

### CODE REVIEW

**Comment by shitz** [comment]
Reviewed 2 of 5 files at r1.
Review status: 2 of 5 files reviewed at latest revision, 10 unresolved discussions.

---

*[go/cert_srv/main.go, line 127 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TFn54cqNpf-Ea-4W5:-L3TFn54cqNpf-Ea-4W6:bn4lp7l) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/main.go#L127)):*
> ```Go
> }
> 
> func reloadConfig(sighup chan os.Signal) {
> ```

This needs a `defer liblog.PanicAndExit()`

---

*[go/cert_srv/conf/conf.go, line 111 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TUqi6ywPZGMOZeUri:-L3TUqi6ywPZGMOZeUrj:b-9w5j5y) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L111)):*
> ```Go
> 		return common.NewBasicError(ErrorStore, err)
> 	}
> 	c.KeyConfLock.Lock()
> ```

As far as I can tell, this doesn't do anything, since nothing else ever tries to acquire this lock. Right now, nothing seems to actually use the `KeyConf`, but if it would be used externally it's dangerous to expose it publicly, since users should always acquire the readlock first.

---

*[go/cert_srv/conf/conf.go, line 115 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3T_NrhIKsdv9U3v-RC:-L3T_NrhIKsdv9U3v-RD:b1gccdc) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L115)):*
> ```Go
> 	keyConf, err := c.loadKeyConf()
> 	if err == nil {
> 		var certVer, trcVer uint64
> ```

Add comment `// Check if new keyconfig is usable with the current TRC and CertChain

---

*[go/cert_srv/conf/conf.go, line 123 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TZUuRUcBKBXYi-TK3:-L3TZUuRUcBKBXYi-TK4:b-suatmb) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L123)):*
> ```Go
> 		}
> 	}
> 	certVer, trcVer, errO := c.checkKeyConf(c.KeyConf)
> ```

Add comment `// Check that old keyconf is usable with the current (possibly new) TRC or CertChain`

---

*[go/cert_srv/conf/conf.go, line 123 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3T_pfBY54QBfiU61OF:-L3T_pfBY54QBfiU61OG:b29mk6n) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L123)):*
> ```Go
> 		}
> 	}
> 	certVer, trcVer, errO := c.checkKeyConf(c.KeyConf)
> ```

`errO` looks weird. I suggest `existingErr` or `nestedErr` or something like that. 

---

*[go/cert_srv/conf/conf.go, line 139 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TW1ygj49LLBeatUyW:-L3TW1ygj49LLBeatUyX:b-5r607q) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L139)):*
> ```Go
> // checkKeyConf checks if key configuration is usable in combination with the active TRC and
> // certificate chain and returns the associated versions.
> func (c *Conf) checkKeyConf(keyConf *trust.KeyConf) (uint64, uint64, error) {
> ```

Seems weird that `checkKeyConf` returns the TRC and CC versions. Where are those even used? Also rename to `verifyKeyConf`

---

*[go/cert_srv/conf/conf.go, line 139 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3Tak1y6J994lqXVoMZ:-L3Tak1y6J994lqXVoM_:bgqo6dv) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L139)):*
> ```Go
> // checkKeyConf checks if key configuration is usable in combination with the active TRC and
> // certificate chain and returns the associated versions.
> func (c *Conf) checkKeyConf(keyConf *trust.KeyConf) (uint64, uint64, error) {
> ```

The program logic is hard to follow (the naming doesnt' help). `verifiyKeyConf` should look roughly like this:
```
func (c *Conf) verifyKeyConf(keyConf *trust.KeyConf) (error) {
    // Get current certificate chain.
    chain := c.Store.GetNewestChain(c.PublicAddr.IA)
    ...
    // Check that the public key in the key config matches the public key in the certificate.
    verKey := common.RawBytes(ed25519.PrivateKey(keyConf.SignKey).Public().(ed25519.PublicKey))
    if !bytes.Equal(chain.Leaf.SubjectSignKey, verKey) {...}
    // Check that the newest certificate chain is verifiable with the current TRC.
    if err := c.verifyCC(chain) ...
    if isCore {
        // Check that the online root key matches the current TRC (or grace)
        ...
        // Check that the offline root key matches the current TRC (or grace)
    }
}
```

---

*[go/cert_srv/conf/conf.go, line 149 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TM81jggVgDTz5alJb:-L3TM81jggVgDTz5alJc:b-t78aex) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L149)):*
> ```Go
> 	}
> 	if usedT == graceT {
> 		log.Warn("KeyConf relys on TRC in grace period", "TRC", usedT)
> ```

'relies'

---

*[go/cert_srv/conf/conf.go, line 164 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TLBxdeYKK0uLqwn1q:-L3TLBxdeYKK0uLqwn1r:b-dldvtj) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L164)):*
> ```Go
> 		return t, graceT, nil
> 	}
> 	onPubKey := ed25519.PrivateKey(keyConf.OnRootKey).Public().(ed25519.PublicKey)
> ```

What speaks against  making `keyConf.OnRootKey` an `ed25519.PrivateKey` during loading? 

---

*[go/lib/trust/store.go, line 70 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TSdev6hmTLoK99v69:-L3TSdew28x9dFYk4ghr:bqel4cf) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/lib/trust/store.go#L70)):*
> ```Go
> }
> 
> // Reload reloads
> ```

Looks like an unfinished docstring.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1377)*

**Comment by Oncilla** [comment]
Review status: 2 of 5 files reviewed at latest revision, 10 unresolved discussions.

---

*[go/cert_srv/conf/conf.go, line 111 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TUqi6ywPZGMOZeUri:-L3TeQLfIis7cW1igzT1:b-321lxp) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L111)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

As far as I can tell, this doesn't do anything, since nothing else ever tries to acquire this lock. Right now, nothing seems to actually use the `KeyConf`, but if it would be used externally it's dangerous to expose it publicly, since users should always acquire the readlock first.
</blockquote></details>

KeyConfig will be used by CS to
- sign certificate reissuance requests
- sign certificate chains (leaf cert and core cert)
- sign TRC updates
etc.

Thus it has to be accessible from main package in cert_srv. I could add wrapper functions in order to guarantee locks being acquired.

---

*[go/cert_srv/conf/conf.go, line 139 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TW1ygj49LLBeatUyW:-L3TfLnyuxNoYcHWRrqN:b3c8roq) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L139)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Seems weird that `checkKeyConf` returns the TRC and CC versions. Where are those even used? Also rename to `verifyKeyConf`
</blockquote></details>

They are used to identify, which cert/TRC versions do authenticate the keys in keyConfig

---

*[go/cert_srv/conf/conf.go, line 139 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3Tak1y6J994lqXVoMZ:-L3TfVXwKG-tIDDci_dW:bidgcko) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L139)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

The program logic is hard to follow (the naming doesnt' help). `verifiyKeyConf` should look roughly like this:
```
func (c *Conf) verifyKeyConf(keyConf *trust.KeyConf) (error) {
    // Get current certificate chain.
    chain := c.Store.GetNewestChain(c.PublicAddr.IA)
    ...
    // Check that the public key in the key config matches the public key in the certificate.
    verKey := common.RawBytes(ed25519.PrivateKey(keyConf.SignKey).Public().(ed25519.PublicKey))
    if !bytes.Equal(chain.Leaf.SubjectSignKey, verKey) {...}
    // Check that the newest certificate chain is verifiable with the current TRC.
    if err := c.verifyCC(chain) ...
    if isCore {
        // Check that the online root key matches the current TRC (or grace)
        ...
        // Check that the offline root key matches the current TRC (or grace)
    }
}
```
</blockquote></details>

`// Check that the newest certificate chain is verifiable with the current TRC.`
`if err := c.verifyCC(chain) ...`

what do you mean by that?
The newest certificate chain might not be verifiable against the current TRC.
Not loading keyConfig in this situation would imply that the CS can not fetch a new version of the certificate chain which is verifiable with the new TRC.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1377)*

**Comment by shitz** [comment]
Review status: 2 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[go/cert_srv/conf/conf.go, line 187 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TpCi20GS5_2Tjjj8V:-L3TpCi20GS5_2Tjjj8W:bfhig03) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L187)):*
> ```Go
> 		return nil, nil, common.NewBasicError("No TRC for own ISD", nil)
> 	}
> 	for ver := t.Version - 1; ver >= 0; ver-- {
> ```

Add a comment for this loop.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1377)*

**Comment by Oncilla** [comment]
Review status: 2 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[go/cert_srv/main.go, line 127 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TFn54cqNpf-Ea-4W5:-L3TvFryvm91pO5Yi50q:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/main.go#L127)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

This needs a `defer liblog.PanicAndExit()`
</blockquote></details>

Done.

---

*[go/cert_srv/conf/conf.go, line 111 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TUqi6ywPZGMOZeUri:-L3YizNWKfcPiujkJzls:bucpf4t) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L111)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

I'm aware that those keys will be used in the future. Wrappers should work fine.
</blockquote></details>

Done.
I omitted offline root key, since CS should not be using it anyway.

---

*[go/cert_srv/conf/conf.go, line 115 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3T_NrhIKsdv9U3v-RC:-L3Tw_ASaIDAqzNQqoL2:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L115)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Add comment `// Check if new keyconfig is usable with the current TRC and CertChain
</blockquote></details>

Done.

---

*[go/cert_srv/conf/conf.go, line 123 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TZUuRUcBKBXYi-TK3:-L3Twa3UHSycWd-qpBSd:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L123)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Add comment `// Check that old keyconf is usable with the current (possibly new) TRC or CertChain`
</blockquote></details>

Done.

---

*[go/cert_srv/conf/conf.go, line 123 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3T_pfBY54QBfiU61OF:-L3TwbBFUYc9Yg-q0X2w:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L123)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

`errO` looks weird. I suggest `existingErr` or `nestedErr` or something like that. 
</blockquote></details>

Done.

---

*[go/cert_srv/conf/conf.go, line 139 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3Tak1y6J994lqXVoMZ:-L3YdLcNQtj-Q2qH02PW:bxtqa38) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L139)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

`verifyCC` would try to verify against the newest TRC or the TRC in the grace period (basically what your code is doing now). My comment should give you a hint for the structure of the method. You'd still need to make sure that you do all the steps that you are currently doing, but just in a different order, with different names, and with some comments to make it easier to follow the code.
</blockquote></details>

This should be a bit cleaner. 

Sidenote: Having the trust store as a DB would make operations a lot easier. (eg. getActiveTRC)
I'm wondering if it makes sense to postpone this PR until trust store actually is DB backed.

---

*[go/cert_srv/conf/conf.go, line 139 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TW1ygj49LLBeatUyW:-L3YduJJm58bE0oL1V1P:b-aqncpg) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L139)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

I don't see them being used in this file?
</blockquote></details>

It was used to set the Conf.TRCVer and Conf.CerVer. I removed it in this commit.
However, when the CS signs anything with the keys in key config, it should know which Cert/TRC can be used for validation.
Having a DB backing the trust store, this is a simple query, but putting it here in Conf is a lot harder to handle and keep track of (especially looking at the changes to come with PR/1376).

---

*[go/cert_srv/conf/conf.go, line 149 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TM81jggVgDTz5alJb:-L3Y_e49NmvaDQ20pKbB:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L149)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

'relies'
</blockquote></details>

Done.

---

*[go/cert_srv/conf/conf.go, line 164 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TLBxdeYKK0uLqwn1q:-L3YjqHZYBdrgU7bkDf7:b-ct83hl) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L164)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

What speaks against  making `keyConf.OnRootKey` an `ed25519.PrivateKey` during loading? 
</blockquote></details>

KeyConf is in the library part of the code base. I think we should not prescribe, which type of keys are stored in it.

---

*[go/cert_srv/conf/conf.go, line 187 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TpCi20GS5_2Tjjj8V:-L3YcVi-wuJWLsK4eRsC:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/cert_srv/conf/conf.go#L187)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Add a comment for this loop.
</blockquote></details>

Done.

---

*[go/lib/trust/store.go, line 70 at r1](https://reviewable.io:443/reviews/scionproto/scion/1377#-L3TSdev6hmTLoK99v69:-L3YdISsfYFl7cR88QUK:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/5836d23c5ae96a6b3c9e690418051494460ff32d/go/lib/trust/store.go#L70)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Looks like an unfinished docstring.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1377)*

**Comment by Oncilla** [comment]
postponed until trust store is DB backed
